feat(normalization): Trim logger smartly

iker-barriocanal opened the pull request:
This PR changes how we trim loggers for events coming from Java platforms. Chars from the end are now prioritized (so the string is trimmed from the beginning), and it's also attempted to trim full words in `.`s.

We are only interested in doing this for Java at this moment; because loggers in Java are fully qualified names (including the package, so potentially very long strings) where the class name at the end is more important, and because it's the only platform we've received this type of feedback on.

Resolves https://github.com/getsentry/relay/issues/1827.

### Examples

These examples are from the tests, see the diff for more examples and/or edge cases.

```txt
# Before
this_is-way_too_long-and_we_only_have_one_word-so_we_cant_smart_trim

# After
*is-way_too_long-and_we_only_have_one_word-so_we_cant_smart_trim
```

```txt
# Before
super_out.this_is_already_out_too.this_part-is-kept.this_right_here-is_a-very_long_word

# After
*.this_part-is-kept.this_right_here-is_a-very_long_word
```

iker-barriocanal committed:
wip

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+        meta: &mut Meta,
+        state: &ProcessingState,
+    ) -> ProcessingResult {
+        event_logger.process_child_values(self, state)?;
```
Why even call `process_child_values`? `EventLogger` will always be a leaf node AFAIK.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+
+static LOGGER_DELIMITER_PATTERN: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"\.").expect("Invalid logger separator Regex"));
+static LOGGER_TRIMMED_PREFIX: &str = "*";
```
```suggestion
const LOGGER_TRIMMED_PREFIX: &str = "*";
```

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+static LOGGER_TRIMMED_PREFIX: &str = "*";
+
+fn shorten_logger(logger: &str) -> String {
+    let mut tokens = LOGGER_DELIMITER_PATTERN.split(logger).collect_vec();
```
Using regex here seems like overkill to me.
```suggestion
    let mut tokens = logger.split(LOGGER_DELIMITER_CHAR).collect_vec();
```

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+    let target_size = logger_target_size(tokens);
+    let target_idx = logger_len - target_size;
+
+    LOGGER_TRIMMED_PREFIX.to_owned() + &logger[target_idx..]
```
`logger[target_idx..]` could panic if `target_idx` is not a character boundary.

E.g. `&"Löwe"[2..]` will result in 

```
thread 'main' panicked at 'byte index 2 is not a char boundary; it is inside 'ö' (bytes 1..3) of `Löwe`'
```

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1001,6 +1002,78 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+
+    for token in tokens {
+        // Consider one slot for the delimiter between tokens.
+        let current = bytecount::num_chars(token.as_bytes()) + 1;
```
It's a good idea to use `num_chars` for limiting the string, but the computed `target_size` that is used for slicing the string should be in bytes. Best add some unit tests with strings that contain a lot of special characters to see the effects.

jjbayer reviewed (commented):
I like the idea of a dedicated `EventLogger` type, _but_ if we want to make this normalization dependent on other fields like `platform`, I think it's easier to add a function to `light_normalize_event` instead, similar to `normalize_logentry`.

iker-barriocanal committed:
simplify and document code

iker-barriocanal committed:
Merge branch 'master' into iker/feat/logger-smart-trimming

iker-barriocanal committed:
Update changelog

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -806,6 +806,19 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+                let trimmed_logger = event_logger.as_str().trim();
+                let logger_len = bytecount::num_chars(trimmed_logger.as_bytes());
+
+                if logger_len > MaxChars::Logger.limit() {
```
You do the same check in `shorten_logger`, so I would remove it here.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -806,6 +806,19 @@ impl<'a> Processor for NormalizeProcessor<'a> {
 
+        if event.platform.as_str() == Some("java") {
+            if let Some(event_logger) = event.logger.value_mut() {
+                let trimmed_logger = event_logger.as_str().trim();
```
Can we move this basic trimming inside `shorten_logger`?

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+///
+/// Additionally, the new logger is prefixed with `*`, to indicate it was
+/// shortened.
+fn shorten_logger(logger: &str) -> String {
```
nit: I would rename this to `trim_logger`, since that is the word we use in other places (e.g. `TrimmingProcessor`).

iker-barriocanal replied:
Unless there're strong opinions about changing that name, I'd rather not. The reason is that what this method does isn't really trimming:

> In computer programming, trimming or stripping is a string manipulation in which leading and trailing whitespace is removed from a string.

We don't trim here. We remove a non-whitespace substring from the string and insert a `*`. I'd like to avoid any future confusion coming from the wording.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+fn shorten_logger(logger: &str) -> String {
+    let logger_len = bytecount::num_chars(logger.as_bytes());
+    if logger_len <= MaxChars::Logger.limit() {
+        return logger.to_owned();
```
You could change the interface of this function to consume a `String`, then you could return `logger` here without extra allocation:

`fn shorten_logger(logger: String) -> String {`

iker-barriocanal replied:
It's a good suggestion. However, I've moved the trimming to this function too and `trim` returns an `str`, so we still have an extra allocation in this case.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+    tokens.pop();
+
+    remove_logger_extra_chars(&mut tokens);
+    remove_logger_word(&mut tokens);
```
Shouldn't this be called only if `remove_logger_extra_chars` cut a word in half? If my logger is `foo.bar`, wouldn't `remove_logger_word` crop it to `.bar`?

iker-barriocanal replied:
That's an interesting case I didn't think about. Should be addressed in db4d016.

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -1005,6 +1018,68 @@ impl<'a> Processor for NormalizeProcessor<'a> {
+    remove_logger_word(&mut tokens);
+
+    tokens.reverse();
+    "*".to_owned() + &tokens.iter().join("")
```
```suggestion
    tokens.join("*")
```

iker-barriocanal replied:
`join` is the contrary operation to `split`. We only want the `*` at the beginning to indicate the logger was trimmed, not between each token (where a token is a character).

jjbayer replied:
Whoops, I guess I meant something like
```suggestion
    format!("*{}", tokens.join(""))
```

iker-barriocanal committed:
Move basic trimming to shorten_logger

iker-barriocanal committed:
Only remove words if cut

If a word has been cut, remove it. If not, don't. This avoids an
edge-case where a word is removed because it has no leading `.`s,
resulting in a too aggressive char removal. See test and snapshot.

iker-barriocanal committed:
Fix lint

jjbayer commented on relay-general/src/store/normalize.rs:
```diff
@@ -808,13 +808,9 @@ impl<'a> Processor for NormalizeProcessor<'a> {
-                    *event_logger = shorten_logger(trimmed_logger);
-                } else if trimmed_logger != event_logger.as_str() {
-                    *event_logger = trimmed_logger.to_owned();
+                let shortened = shorten_logger(event_logger.clone());
```
nit: It must be possible to prevent this clone. Maybe change the signature to `shorten_logger(&mut String) -> ()`? Then the check below becomes unnecessary.

iker-barriocanal replied:
If we change the signature to accept a `&mut String`, we remove a cloning here. However, in cases where the logger is smaller than the max size, we will always perform an allocation (if we clone the String, not always) which is not the case for cloning; plus, making the code less readable (why do we need to pass a mutable string?). So optimizing to remove extra allocations doesn't always happen.

I don't have data to back up the following, but my guess is that the performance improvement is not noticeable. This makes me optimize how the code is used instead.

Of course, I may be missing something that makes this possible _and_ readable. If that's the case, I'm happy to update the code. If not, I'll go ahead and merge it like this.

jjbayer replied:
> I may be missing something that makes this possible and readable.

Opened a little draft PR, let me know what you think: https://github.com/getsentry/relay/pull/1950

iker-barriocanal committed:
Format string with format!

iker-barriocanal committed:
Remove extra allocation

iker-barriocanal committed:
Add meta remarks on logger modification

iker-barriocanal committed:
Merge branch 'master' into iker/feat/logger-smart-trimming
